kafka/src/main/java/com/metamx/tranquility/kafka/KafkaConsumer.java

```diff
@@ -0,0 +1,251 @@
+import java.util.concurrent.Executors;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+/**
+ * Spawns a number of threads to read messages from Kafka topics and write them by calling
+ * WriterController.getWriter(topic).send(). Will periodically call WriterController.flushAll() and when this completes
+ * will call ConsumerConnector.commitOffsets() to save the last written offset to ZooKeeper. This implementation
+ * guarantees that any events in Kafka will be read at least once even in case of a failure condition but does not
+ * guarantee that duplication will not occur.
+ */
+public class KafkaConsumer
+{
+  private static final Logger log = LoggerFactory.getLogger(KafkaConsumer.class);
+
+  private final ExecutorService consumerExec;
+  private final Thread commitThread;
+  private final AtomicBoolean shutdown = new AtomicBoolean();
+
+  // prevents reading the next event from Kafka while events are being flushed and offset is being committed to ZK
+  private final ReentrantReadWriteLock commitLock = new ReentrantReadWriteLock();
+
+  private final ConsumerConnector consumerConnector;
+  private final TopicFilter topicFilter;
+  private final int numThreads;
+  private final int commitMillis;
+  private final WriterController writerController;
+
+  public KafkaConsumer(
+      final Properties kafkaProperties,
+      final String schemaMap,
+      final int numThreads,
+      final int commitMillis,
+      final WriterController writerController
+  )
+  {
+    this.consumerConnector = getConsumerConnector(kafkaProperties);
+    this.topicFilter = new Whitelist(buildTopicFilter(schemaMap));
+    this.numThreads = numThreads;
+    this.commitMillis = commitMillis;
+    this.writerController = writerController;
+
+    this.consumerExec = Executors.newFixedThreadPool(numThreads,
+                                 new ThreadFactoryBuilder().setNameFormat("KafkaConsumer-%d").setDaemon(true).build());
+
+    this.commitThread = new Thread(createCommitRunnable());
+    this.commitThread.setName("KafkaConsumer-CommitThread");
+    this.commitThread.setDaemon(true);
+  }
+
+  public void start()
+  {
+    commitThread.start();
+    startConsumers();
+  }
+
+  public void stop()
+  {
+    if (shutdown.compareAndSet(false, true)) {
+      log.info("Shutting down - attempting to flush buffers and commit final offsets");
+
+      try {
+        commitLock.writeLock().lockInterruptibly(); // prevent Kafka from consuming any more events
+        writerController.stop(); // try to flush the remaining events to Druid
+        writerController.waitForAll(); // wait for all writers to finish flushing before committing offsets
+        consumerConnector.commitOffsets(); // update commit offset
+      } catch (InterruptedException e) {
+        Throwables.propagate(e);
+      } finally {
+        consumerConnector.shutdown();
+        commitLock.writeLock().unlock();
```

Comment:
Agreed, it's good to move `commitLock.writeLock().unlock()` first and only call it if we've acquired the lock.